design/open/hard-delete.md

```diff
@@ -0,0 +1,202 @@
+
+#### LinkPhysicalAddress
+
+Assume this API uses a physical address in the repo namespace
+1. Read object's metadata
+2. Increment counter and write metadata
+3. Add staging entry
+
+#### StageObject
+
+If physical address in repo namespace - perform increment as described above, otherwise do not handle retention
+
+#### Reset Branch
+
+When resetting a branch we throw all the uncommitted branch data.
+We can leverage the new KV drop async functionality to also check and hard-delete objects as needed
+
+### Atomic reference counter
+
+For each storage adapter (AWS, GC, AZ), we will use a builtin logic to provide this functionality
+
+#### AWS
+
+AWS doesn't allow to update an object or its metadata after creation. In order to update the metadata a copy object should be performed with the new metadata values. The only 
+conditionals currently available are on the objects timestamp and ETag. Unfortunately ETags are modified only on objects content change and does not take into account changes in metadata.
+
+#### Azure
+
+Store the reference counter in the blob's metadata and use `set blob metadata` API with the supported conditional header on the ETag to perform an If-Match operation and update
+the reference counter
+
+
+#### Google Store
+
+Store the reference counter in the blob's metadata, use the `patch object` API with the `ifMetagenerationMatch` conditional.
+
+
+## 3. Tracking references in staging manager
+
+Staging manager will track physical addresses of staged objects and the staging tokens referencing them.
+We will introduce a new object to the database:
+
+      key    = <repo>/references/<physical_address>/<staging_token>  
+      value  = state enum (staged|deleted|comitted)
+
+On Upload object to a branch, we will add a key in the references path with the physical address and current staging token and 
+mark it as staged.   
+On Delete object from a staging area, we will update the reference key with value `deleted`  
+On Commit, we will update the reference key with value `comitted`
+
+A background job will be responsible for scanning the references prefix, and handling the references according to the state.
+
+
+### Background delete job (pseudo-code)
+
+1. Scan `reference` prefix (can use after prefix for load balancing)
+2. For each `physical_address` read all entries
+      1. If found state == 'committed' in any entry
+         1. Delete all keys for `phisycal_address`, by order of state: deleted -> staged -> committed
+      2. If state == 'deleted' in all entries
+         1. Hard-delete object
+         2. Delete all keys for `phisycal_address`
+
+### Key Update Order of Precedence
+
+1. `commited` state takes precedence over all (terminal state - overrides any other value) - uses **Set**
+2. `deleted` state can only be done on entries which are not `committed` and uses **SetIf**
+3. `staged` state can only be done on entries which are not `committed` and uses **SetIf** 
+
+### Flows
```

Comment:
How about branch deletion, when committed objects are left unreferenced?
Does the ref count cover it too?

Reply:
Yes, on branch deletion, and as part of the staging token drop, we scan over token entries,
if entry was tombstone - we delete the reference entry, otherwise we modify the reference state to deleted

Reply:
This is the same as reset branch

Reply:
I'm not sure I follow on this one - I'm asking about a branch with all data commited - no objects at staging are for that matter, and say a certain object is only referenced by this branch. What happens when the branch is deleted? The (now) unreferenced object is not accessible via any of staging tokens, as these are gone

Reply:
We do not track committed data references, once data is committed the reference keys will eventually be deleted (only the reference keys, not the data) by the background process. Data which is committed is referenced by its commit, not its branch. By design we do not deleted these objects. This proposal discusses only the garbage collection on uncommitted data. For committed data we have the external GC process retention policy. I hope I answered what was asked

Reply:
Actually you did. Thanks